textile/features.textile

```diff
@@ -1510,43 +1508,52 @@ h4. ChannelMetrics
-* @(BPD2)@ @BatchPresenceResponse@ contains the following attributes:
-** @(BPD2a)@ @channel@ is the channel name which this presence request
-** @(BPD2b)@ @presence@ is an array of presence data for the @channel@
-
-h4. BatchPresence
-* @(BPE1)@ Is a partial @PresenceMessage@ object containing @clientId@ and @action@, or @error@ if the presence failed
-* @(PBE2)@ This clause has been renamed to #BPE2.
-** @(PBE2a)@ This clause has been renamed to #BPE2a.
-** @(PBE2b)@ This clause has been renamed to #BPE2b.
-** @(PBE2c)@ This clause has been renamed to #BPE2c.
-* @(BPE2)@ @BatchPresence@ contains the following attributes:
-** @(BPE2a)@ @clientId@ - identical to #TP3c
-** @(BPE2b)@ @action@ - identical to #TP3b - null if @error@ is present
-** @(BPE2c)@ @error@ - an @ErrorInfo@ object representing the failure reason for this channel - null if @action@ is present
+h4. BatchPublishResult
+
+* @(BPR1)@ Contains information about the result of publishes requested by a single @BatchPublishSpec@
+* @(BPR2)@ The attributes of @BatchPublishResult@ consist of:
+** @(BPR2a)@ @channel@ string - the name of the channel
+** @(BPR2b)@ @messageId@ string - an optional string containing the @messageId@ prefix for the published message(s), if successful
+** @(BPR2c)@ @error@ @ErrorInfo@ - an optional @ErrorInfo@ indicating the reason the message(s) failed to publish, if unsuccessful
+
+h4. BatchPresenceResult
+
+* @(BGR1)@ Contains information about the result of a batch presence request for a single channel
+* @(BGR2)@ The attributes of @BatchPublishChannelResult@ consist of:
+** @(BGR2a)@ @channel@ string - the name of the channel
+** @(BGR2b)@ @presence@ @PresenceMessage[]@ - an optional array containing all members present on the channel if the request succeeded
+** @(BGR2c)@ @error@ @ErrorInfo@ - an optional @ErrorInfo@ indicating the reason the presence request failed for the given channel, if unsuccessful
+
+h4. TokenRevocationSpecifierTarget
+
+* @(TRT1)@ Describes which tokens should be affected by a token revocation request
+* @(TRT2)@ The attributes of @TokenRevocationTargetSpecifier@ consist of:
+** @(TRT2a)@ @type@ string - either "clientId" or "revocationKey"
+** @(TRT2b)@ @value@ string - the value of the token revocation target specifier
+
+h4. TokenRevocationResult
+
+* @(TRR1)@ Contains information about the result of a token revocation request for a single target specifier
+* @(TRR2)@ The attributes of @TokenRevocationTargetResult@ consist of:
+** @(TRR2a)@ @target@ string - the target specifier
+** @(TRR2b)@ @appliesAt@ number - an optional timestamp at which the token revocation will take effect, if successful
+** @(TRR2c)@ @issuedBefore@ number - an optional timestamp for which tokens previously issued will be revoked, if successful
+** @(TRR2d)@ @error@ @ErrorInfo@ - an optional @ErrorInfo@ indicating the reason the token revocation request failed for the given specified, if unsuccessful
```

Comment:
wonder if we could be noting a slightly stronger type here? Like, the server guarantees that a targetresult will be `{ target, issuedBefore, appliesAt } | { target, error }`, would be nice to encode that guarantee in the spec rather than just marking every field other than `target` as optional, for sdks that can represent that in the type system. Not sure what the best way of representing that here would be though

Reply:
I've changed it so that each 'result' type has a success and a failure variant so that this constraint is expressed correctly.

Reply:
nice :+1: